src/command/expire/expire.c

```diff
@@ -863,6 +863,88 @@ removeExpiredBackup(InfoBackup *infoBackup, const String *adhocBackupLabel, unsi
+            String *minBackupLabel = strNewFmt("%sF", buffer);
+
+            // Get all the history years.
+            const StringList *historyYearList = strLstSort(storageListP(
+                storageRepo(), STRDEF(STORAGE_REPO_BACKUP "/" BACKUP_PATH_HISTORY), .expression = STRDEF("^2[0-9]{3}$")),
+                sortOrderDesc);
+            
+            for (unsigned int historyYearIdx = 0; historyYearIdx < strLstSize(historyYearList); historyYearIdx++)
+            {
+                // Get all the backup history manifests.
+                const String *historyYear = strLstGet(historyYearList, historyYearIdx);
+                const StringList *historyList = strLstSort(storageListP(
+                    storageRepo(), strNewFmt(STORAGE_REPO_BACKUP "/" BACKUP_PATH_HISTORY "/%s", strZ(historyYear)),
+                        .expression = strNewFmt(
+                            "%s\\.manifest\\.%s$",
+                            strZ(backupRegExpP(.full = true, .differential = true, .incremental = true, .noAnchorEnd = true)),
+                            strZ(compressTypeStr(compressTypeGz)))),
+                    sortOrderDesc);
+
+                for (unsigned int historyIdx = 0; historyIdx < strLstSize(historyList); historyIdx++)
+                {
+                    const String *historyBackupFile = strLstGet(historyList, historyIdx);
+                    const String *historyBackupLabel = strLstGet(strLstNewSplitZ(historyBackupFile, "."), 0);
+
+                    // Keep the history manifests for unexpired backups and compare dates, extracted from backup labels since the 
+                    // retention is set in days.
+                    if (!strLstExists(currentBackupList, historyBackupLabel) &&
+                        strCmp(strSubN(historyBackupLabel, 0, 8), strSubN(minBackupLabel, 0, 8)) < 0)
+                    {
+                        LOG_INFO_FMT(
+                        "repo%u: remove expired history backup manifest %s", cfgOptionGroupIdxToKey(cfgOptGrpRepo, repoIdx),
+                        strZ(historyBackupFile));
+
+                        // Execute the real expiration and deletion only if the dry-run mode is disabled.
+                        if (!cfgOptionValid(cfgOptDryRun) || !cfgOptionBool(cfgOptDryRun))
+                        {
+                            storageRemoveP(
+                                storageRepoIdxWrite(repoIdx),
+                                strNewFmt(
+                                    STORAGE_REPO_BACKUP "/" BACKUP_PATH_HISTORY "/%s/%s",
+                                    strZ(historyYear), strZ(historyBackupFile)));
+                        }
+                    }
+                }
+
+                // Now that we removed the history manifests, we could remove the year directory if empty. However, it seems better
+                // to leave it there (and empty) to avoid the extra callback to the storage.                 
```

Comment:
I think we should remove the year directory once it is empty. No extra call should be required if you track the number of files removed in the loop above.

But we should never remove the `backup.history` directory itself.

Reply:
I thought about it too but that's if we **only** have `x.manifest.x` files in it (matching the storage expression). Else, the `storagePathRemoveP` function would ERROR returning `[39] Directory not empty`.

And I'm not really optimistic to use `.recurse = true` which would remove all the files there in case user would put some files there manually...

So:
- or we leave the directory there (even if empty)
- or we add an extra callback to really check if empty
- or we remove with `.recurse = true`
- or we let the ERROR happen and recommend the users to clean it manually

I was in favor of (1) :-) What's your thoughts about this?

Reply:
@dwsteele Correct me if I'm wrong, but won't we create the directory each time we want to store a manifest in the history directory and the directory does not exist? If so, then if the directory is empty after expiring the files, we should remove it since it will be automatically created the next time if it is needed.

Reply:
The directory is created if missing (given a quick test I just ran). So, d799a7a adds the directory removal using `.recurse = true`. I've not updated the tests yet, just waiting your advice on this commit before doing so.

I list the *manifest* files in the year directory, count the number of removal and remove the directory if removed all the *manifest* found.

My concern is that users would add other files in that directory manually. The only way to make sure that the directory is empty is to trigger another look on disk without filtering any file expression. Perhaps a side note in the doc like **"When all history manifests files are expired, the year directory will be expired too."** (so, users know they probably shouldn't put anything else in those directories...)

Reply:
I have little sympathy for users placing their own files in the repository. It's certainly not a good reason for us to not clean up after ourselves.

In archive expiration, we simply remove the old directories regardless of the contents. The same is true for backups, where we remove the directory without checking for user-added files. I'm in favor of the same behavior here, i.e. use `recurse`.

This would be easier if we did it the way archive expiration does. If the entire year is less than the year of the minimum backup to retain, then just remove the dir (see https://github.com/pgbackrest/pgbackrest/blob/release/2.33/src/command/expire/expire.c#L656). Otherwise find and remove individual files.

Adding a note about this to the option reference is also fine, of course.

Reply:
Hm, you're right, that'd be totally better. And the behavior would be consistent with other expires.
I've then changed the year directory removal and added the tests in 9886532.
Thanks for your inputs!

Reply:
That looks good!